crates/aptos-substreams/src/transaction_converter.rs

```diff
@@ -0,0 +1,275 @@
+            Ed25519Signature, Event, MultiAgentSignature, MultiEd25519Signature, TransactionInfo,
+            UserTransaction, UserTransactionRequest,
+        },
+        block_output::{self, EventKeyOutput, SignatureOutput},
+    },
+};
+use anyhow::{bail, Context, Result};
+use block_output::{
+    BlockMetadataTransactionOutput, EventOutput, TransactionInfoOutput, UserTransactionOutput,
+    WriteSetChangeOutput,
+};
+
+pub fn get_transaction_info_output(
+    txn: &Transaction,
+    info: &TransactionInfo,
+) -> Result<TransactionInfoOutput> {
+    let transaction_info = TransactionInfoOutput {
+        hash: info.hash.clone(),
+        r#type: get_transaction_type(txn.r#type()),
+        version: txn.version,
+        state_root_hash: info.state_root_hash.clone(),
+        event_root_hash: info.event_root_hash.clone(),
+        gas_used: info.gas_used,
+        success: info.success,
+        epoch: txn.epoch,
+        block_height: txn.block_height,
+        vm_status: info.vm_status.clone(),
+        accumulator_root_hash: info.accumulator_root_hash.clone(),
+        timestamp: txn.timestamp.clone(),
+    };
+    Ok(transaction_info)
+}
+
+pub fn get_block_metadata_output(
+    bmt: &BlockMetadataTransaction,
+    info: &TransactionInfoOutput,
+) -> Result<BlockMetadataTransactionOutput> {
+    let bmt_output = BlockMetadataTransactionOutput {
+        hash: info.hash.clone(),
+        id: bmt.id.clone(),
+        round: bmt.round,
+        previous_block_votes_bitmap: bmt.previous_block_votes.clone(),
+        proposer: bmt.proposer.clone(),
+        failed_proposer_indices: bmt.failed_proposer_indices.clone(),
+        timestamp: info.timestamp.clone(),
+        epoch: info.epoch,
+    };
+    Ok(bmt_output)
+}
+
+pub fn get_user_transaction_output(
+    user_txn: &UserTransaction,
+    info: &TransactionInfoOutput,
+) -> Result<UserTransactionOutput> {
+    if let Some(user_request) = &user_txn.request {
+        let mut signature_type = String::new();
+        let mut signatures = vec![];
+        if let Some(signature) = &user_request.signature {
+            signature_type = get_signature_type(signature.r#type());
+            if let Some(signature) = &signature.signature {
+                signatures = get_signature_outputs(signature, user_request, info)?;
+            }
+        }
+        let user_txn_output = UserTransactionOutput {
+            hash: info.hash.clone(),
+            sender: user_request.sender.clone(),
+            sequence_number: user_request.sequence_number,
+            max_gas_amount: user_request.max_gas_amount,
+            expiration_timestamp_secs: user_request.expiration_timestamp_secs.clone(),
+            gas_unit_price: user_request.gas_unit_price,
+            timestamp: info.timestamp.clone(),
+            parent_signature_type: signature_type,
+            signatures,
+        };
+        Ok(user_txn_output)
+    } else {
+        bail!("Transaction info missing from Transaction")
```

Comment:
if a user transaction is malformed, we want to panic; otherwise we'll easily miss logs, end up with bad data, the remedy for which could be rather painful